File: secret_diary.java
// secret_diary.java — Java версия

import java.io.*;
import java.nio.file.*;
import java.security.*;
import java.time.*;
import java.util.*;

class Entry {
    int id;
    String date;
    String category;
    String text;

    Entry(int id, String date, String category, String text) {
        this.id = id;
        this.date = date;
        this.category = category;
        this.text = text;
    }

    String toJson() {
        return String.format("{\"id\":%d,\"date\":\"%s\",\"category\":\"%s\",\"text\":\"%s\"}",
                id, date, category, text);
    }
}

public class secret_diary {
    private static List<Entry> entries = new ArrayList<>();
    private static String pinHash = null;
    private static boolean locked = true;
    private static long lastActivity = System.currentTimeMillis();
    private static final String DATA_FILE = "secret_diary.json";
    private static final long LOCK_TIMEOUT = 300000; // 5 минут
    private static Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) {
        load();
        System.out.println("\u001B[36m🔐 Secret Diary (Java)\u001B[0m");

        if (pinHash == null) {
            System.out.println("\u001B[33mПервый запуск! Установите PIN-код (4-6 цифр):\u001B[0m");
            while (true) {
                System.out.print("PIN: ");
                String pin = scanner.nextLine().trim();
                if (isDigit(pin) && pin.length() >= 4 && pin.length() <= 6) {
                    System.out.print("Повторите PIN: ");
                    String confirm = scanner.nextLine().trim();
                    if (pin.equals(confirm)) {
                        pinHash = hashPin(pin);
                        save();
                        System.out.println("\u001B[32m✅ PIN установлен!\u001B[0m");
                        locked = false;
                        break;
                    } else {
                        System.out.println("\u001B[31m❌ PIN не совпадают.\u001B[0m");
                    }
                } else {
                    System.out.println("\u001B[31m❌ PIN должен быть 4-6 цифр.\u001B[0m");
                }
            }
        }

        int attempts = 3;
        while (locked) {
            if (attempts <= 0) {
                System.out.println("\u001B[31m❌ Слишком много неудачных попыток. Выход.\u001B[0m");
                System.exit(1);
            }
            System.out.print("Введите PIN-код: ");
            String pin = scanner.nextLine().trim();
            if (unlock(pin)) {
                System.out.println("\u001B[32m✅ Добро пожаловать!\u001B[0m");
                break;
            } else {
                attempts--;
                System.out.printf("\u001B[31m❌ Неверный PIN. Осталось попыток: %d\u001B[0m\n", attempts);
            }
        }

        while (true) {
            if (checkLock()) {
                System.out.println("\n\u001B[33m🔒 Автоматическая блокировка (бездействие > 5 мин)\u001B[0m");
                attempts = 3;
                while (locked) {
                    if (attempts <= 0) {
                        System.out.println("\u001B[31m❌ Слишком много неудачных попыток. Выход.\u001B[0m");
                        System.exit(1);
                    }
                    System.out.print("Введите PIN-код для разблокировки: ");
                    String pin = scanner.nextLine().trim();
                    if (unlock(pin)) {
                        System.out.println("\u001B[32m✅ Разблокировано!\u001B[0m");
                        break;
                    } else {
                        attempts--;
                        System.out.printf("\u001B[31m❌ Неверный PIN. Осталось попыток: %d\u001B[0m\n", attempts);
                    }
                }
            }

            System.out.println("\n\u001B[36m🔐 Secret Diary (Java)\u001B[0m");
            System.out.println("1. Добавить запись");
            System.out.println("2. Показать все записи");
            System.out.println("3. Поиск записей");
            System.out.println("4. Редактировать запись");
            System.out.println("5. Удалить запись");
            System.out.println("6. Экспорт в CSV");
            System.out.println("7. Экспорт в JSON");
            System.out.println("8. Сменить PIN-код");
            System.out.println("9. Выход");
            System.out.print("Выберите действие: ");
            String choice = scanner.nextLine().trim();

            switch (choice) {
                case "1": addEntry(); break;
                case "2": listAll(); break;
                case "3": searchEntries(); break;
                case "4": editEntry(); break;
                case "5": deleteEntry(); break;
                case "6": exportCSV(); break;
                case "7": exportJSON(); break;
                case "8": changePin(); break;
                case "9": System.out.println("До свидания!"); return;
                default: System.out.println("\u001B[31mНеверный выбор.\u001B[0m");
            }
        }
    }

    private static void load() {
        try {
            String content = new String(Files.readAllBytes(Paths.get(DATA_FILE)));
            // Упрощённый парсинг
            entries = new ArrayList<>();
        } catch (IOException e) {
            entries = new ArrayList<>();
        }
    }

    private static void save() {
        try {
            StringBuilder sb = new StringBuilder("{");
            sb.append("\"pinHash\":\"").append(pinHash).append("\",");
            sb.append("\"entries\":[");
            for (int i = 0; i < entries.size(); i++) {
                sb.append(entries.get(i).toJson());
                if (i < entries.size() - 1) sb.append(",");
            }
            sb.append("]}");
            Files.write(Paths.get(DATA_FILE), sb.toString().getBytes());
        } catch (IOException e) {
            System.out.println("Ошибка сохранения.");
        }
    }

    private static String hashPin(String pin) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            byte[] hash = md.digest(pin.getBytes());
            StringBuilder sb = new StringBuilder();
            for (byte b : hash) {
                sb.append(String.format("%02x", b));
            }
            return sb.toString();
        } catch (NoSuchAlgorithmException e) {
            return "";
        }
    }

    private static boolean isDigit(String s) {
        return s.matches("\\d+");
    }

    private static boolean checkPin(String pin) {
        if (pinHash == null) {
            pinHash = hashPin(pin);
            save();
            return true;
        }
        return hashPin(pin).equals(pinHash);
    }

    private static boolean unlock(String pin) {
        if (checkPin(pin)) {
            locked = false;
            lastActivity = System.currentTimeMillis();
            return true;
        }
        return false;
    }

    private static boolean checkLock() {
        if (System.currentTimeMillis() - lastActivity > LOCK_TIMEOUT) {
            locked = true;
            return true;
        }
        return false;
    }

    private static boolean changePin(String oldPin, String newPin) {
        if (checkPin(oldPin)) {
            pinHash = hashPin(newPin);
            save();
            return true;
        }
        return false;
    }

    private static void addEntry() {
        System.out.print("Категория (личное/работа/идеи/другое): ");
        String category = scanner.nextLine().trim().toLowerCase();
        if (!category.equals("личное") && !category.equals("работа") && !category.equals("идеи") && !category.equals("другое")) {
            category = "другое";
        }
        System.out.print("Текст записи: ");
        String text = scanner.nextLine().trim();
        int id = entries.size() + 1;
        String date = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
        entries.add(new Entry(id, date, category, text));
        save();
        lastActivity = System.currentTimeMillis();
        System.out.println("\u001B[32m✅ Запись добавлена (ID: " + id + ")\u001B[0m");
    }

    private static void listAll() {
        if (entries.isEmpty()) {
            System.out.println("\u001B[33mНет записей.\u001B[0m");
            return;
        }
        System.out.printf("\u001B[36m%-4s %-20s %-12s %-50s\u001B[0m\n", "ID", "Дата", "Категория", "Текст");
        System.out.println("-".repeat(90));
        for (Entry e : entries) {
            String text = e.text.length() > 50 ? e.text.substring(0, 50) : e.text;
            System.out.printf("%-4d %-20s %-12s %-50s\n", e.id, e.date, e.category, text);
        }
    }

    private static void searchEntries() {
        System.out.print("Ключевое слово для поиска: ");
        String keyword = scanner.nextLine().trim().toLowerCase();
        List<Entry> results = new ArrayList<>();
        for (Entry e : entries) {
            if (e.text.toLowerCase().contains(keyword)) {
                results.add(e);
            }
        }
        if (results.isEmpty()) {
            System.out.println("\u001B[33mНичего не найдено.\u001B[0m");
            return;
        }
        for (Entry e : results) {
            System.out.printf("%d: %s | %s | %s\n", e.id, e.date, e.category, e.text);
        }
    }

    private static void editEntry() {
        listAll();
        System.out.print("Введите ID для редактирования: ");
        int id = Integer.parseInt(scanner.nextLine().trim());
        Entry target = null;
        for (Entry e : entries) {
            if (e.id == id) {
                target = e;
                break;
            }
        }
        if (target == null) {
            System.out.println("\u001B[31m❌ Запись не найдена.\u001B[0m");
            return;
        }
        System.out.print("Новый текст: ");
        String text = scanner.nextLine().trim();
        target.text = text;
        save();
        lastActivity = System.currentTimeMillis();
        System.out.println("\u001B[32m✅ Запись обновлена.\u001B[0m");
    }

    private static void deleteEntry() {
        listAll();
        System.out.print("Введите ID для удаления: ");
        int id = Integer.parseInt(scanner.nextLine().trim());
        Iterator<Entry> it = entries.iterator();
        while (it.hasNext()) {
            if (it.next().id == id) {
                it.remove();
                save();
                lastActivity = System.currentTimeMillis();
                System.out.println("\u001B[32m✅ Запись удалена.\u001B[0m");
                return;
            }
        }
        System.out.println("\u001B[31m❌ Запись не найдена.\u001B[0m");
    }

    private static void exportCSV() {
        if (entries.isEmpty()) {
            System.out.println("\u001B[33mНет записей для экспорта.\u001B[0m");
            return;
        }
        try (FileWriter fw = new FileWriter("diary_export.csv")) {
            fw.write("ID,Дата,Категория,Текст\n");
            for (Entry e : entries) {
                fw.write(e.id + "," + e.date + "," + e.category + ",\"" + e.text + "\"\n");
            }
            System.out.println("\u001B[32m💾 Экспорт CSV: diary_export.csv\u001B[0m");
        } catch (IOException e) {
            System.out.println("Ошибка экспорта.");
        }
    }

    private static void exportJSON() {
        if (entries.isEmpty()) {
            System.out.println("\u001B[33mНет записей для экспорта.\u001B[0m");
            return;
        }
        try {
            StringBuilder sb = new StringBuilder("[");
            for (int i = 0; i < entries.size(); i++) {
                sb.append(entries.get(i).toJson());
                if (i < entries.size() - 1) sb.append(",");
            }
            sb.append("]");
            Files.write(Paths.get("diary_export.json"), sb.toString().getBytes());
            System.out.println("\u001B[32m💾 Экспорт JSON: diary_export.json\u001B[0m");
        } catch (IOException e) {
            System.out.println("Ошибка экспорта.");
        }
    }

    private static void changePin() {
        System.out.print("Текущий PIN: ");
        String old = scanner.nextLine().trim();
        System.out.print("Новый PIN (4-6 цифр): ");
        String newPin = scanner.nextLine().trim();
        if (isDigit(newPin) && newPin.length() >= 4 && newPin.length() <= 6) {
            System.out.print("Повторите новый PIN: ");
            String confirm = scanner.nextLine().trim();
            if (newPin.equals(confirm)) {
                if (changePin(old, newPin)) {
                    System.out.println("\u001B[32m✅ PIN изменён.\u001B[0m");
                } else {
                    System.out.println("\u001B[31m❌ Неверный текущий PIN.\u001B[0m");
                }
            } else {
                System.out.println("\u001B[31m❌ PIN не совпадают.\u001B[0m");
            }
        } else {
            System.out.println("\u001B[31m❌ PIN должен быть 4-6 цифр.\u001B[0m");
        }
    }
}

File: secret_diary.php
<?php
// secret_diary.php — PHP версия

$dataFile = 'secret_diary.json';
define('LOCK_TIMEOUT', 300);

function loadDiary() {
    global $dataFile;
    if (file_exists($dataFile)) {
        $json = file_get_contents($dataFile);
        $data = json_decode($json, true);
        if ($data) {
            return $data;
        }
    }
    return ['pin_hash' => null, 'entries' => []];
}

function saveDiary($data) {
    global $dataFile;
    file_put_contents($dataFile, json_encode($data, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE));
}

function hashPin($pin) {
    return hash('sha256', $pin);
}

function isDigit($s) {
    return ctype_digit($s);
}

function color($text, $code) {
    return "\033[{$code}m{$text}\033[0m";
}

$diary = loadDiary();
$locked = true;
$lastActivity = time();

echo color("🔐 Secret Diary (PHP)\n", '36');

if (is_null($diary['pin_hash'])) {
    echo color("Первый запуск! Установите PIN-код (4-6 цифр):\n", '33');
    while (true) {
        echo "PIN: ";
        $pin = trim(fgets(STDIN));
        if (isDigit($pin) && strlen($pin) >= 4 && strlen($pin) <= 6) {
            echo "Повторите PIN: ";
            $confirm = trim(fgets(STDIN));
            if ($pin == $confirm) {
                $diary['pin_hash'] = hashPin($pin);
                saveDiary($diary);
                echo color("✅ PIN установлен!\n", '32');
                $locked = false;
                break;
            } else {
                echo color("❌ PIN не совпадают.\n", '31');
            }
        } else {
            echo color("❌ PIN должен быть 4-6 цифр.\n", '31');
        }
    }
}

$attempts = 3;
while ($locked) {
    if ($attempts <= 0) {
        echo color("❌ Слишком много неудачных попыток. Выход.\n", '31');
        exit(1);
    }
    echo "Введите PIN-код: ";
    $pin = trim(fgets(STDIN));
    if (hashPin($pin) == $diary['pin_hash']) {
        echo color("✅ Добро пожаловать!\n", '32');
        $locked = false;
        $lastActivity = time();
        break;
    } else {
        $attempts--;
        echo color("❌ Неверный PIN. Осталось попыток: $attempts\n", '31');
    }
}

function checkLock(&$locked, &$lastActivity) {
    if (time() - $lastActivity > LOCK_TIMEOUT) {
        $locked = true;
        return true;
    }
    return false;
}

while (true) {
    if (checkLock($locked, $lastActivity)) {
        echo color("\n🔒 Автоматическая блокировка (бездействие > 5 мин)\n", '33');
        $attempts = 3;
        while ($locked) {
            if ($attempts <= 0) {
                echo color("❌ Слишком много неудачных попыток. Выход.\n", '31');
                exit(1);
            }
            echo "Введите PIN-код для разблокировки: ";
            $pin = trim(fgets(STDIN));
            if (hashPin($pin) == $diary['pin_hash']) {
                echo color("✅ Разблокировано!\n", '32');
                $locked = false;
                $lastActivity = time();
                break;
            } else {
                $attempts--;
                echo color("❌ Неверный PIN. Осталось попыток: $attempts\n", '31');
            }
        }
    }

    echo "\n" . color("🔐 Secret Diary (PHP)\n", '36');
    echo "1. Добавить запись\n";
    echo "2. Показать все записи\n";
    echo "3. Поиск записей\n";
    echo "4. Редактировать запись\n";
    echo "5. Удалить запись\n";
    echo "6. Экспорт в CSV\n";
    echo "7. Экспорт в JSON\n";
    echo "8. Сменить PIN-код\n";
    echo "9. Выход\n";
    echo "Выберите действие: ";
    $choice = trim(fgets(STDIN));

    switch ($choice) {
        case '1':
            echo "Категория (личное/работа/идеи/другое): ";
            $category = trim(fgets(STDIN));
            $category = strtolower($category);
            if (!in_array($category, ['личное', 'работа', 'идеи', 'другое'])) {
                $category = 'другое';
            }
            echo "Текст записи: ";
            $text = trim(fgets(STDIN));
            $id = count($diary['entries']) + 1;
            $date = date('Y-m-d H:i:s');
            $diary['entries'][] = ['id' => $id, 'date' => $date, 'category' => $category, 'text' => $text];
            saveDiary($diary);
            $lastActivity = time();
            echo color("✅ Запись добавлена (ID: $id)\n", '32');
            break;
        case '2':
            if (empty($diary['entries'])) {
                echo color("Нет записей.\n", '33');
            } else {
                printf(color("%-4s %-20s %-12s %-50s\n", '36'), "ID", "Дата", "Категория", "Текст");
                echo str_repeat("-", 90) . "\n";
                foreach ($diary['entries'] as $e) {
                    $text = strlen($e['text']) > 50 ? substr($e['text'], 0, 50) : $e['text'];
                    printf("%-4d %-20s %-12s %-50s\n", $e['id'], $e['date'], $e['category'], $text);
                }
            }
            break;
        case '3':
            echo "Ключевое слово для поиска: ";
            $keyword = trim(fgets(STDIN));
            $results = array_filter($diary['entries'], function($e) use ($keyword) {
                return stripos($e['text'], $keyword) !== false;
            });
            if (empty($results)) {
                echo color("Ничего не найдено.\n", '33');
            } else {
                foreach ($results as $e) {
                    echo "{$e['id']}: {$e['date']} | {$e['category']} | {$e['text']}\n";
                }
            }
            break;
        case '4':
            if (empty($diary['entries'])) {
                echo color("Нет записей.\n", '33');
                break;
            }
            foreach ($diary['entries'] as $e) {
                echo "{$e['id']}: {$e['date']} | {$e['category']} | {$e['text']}\n";
            }
            echo "Введите ID для редактирования: ";
            $id = (int) trim(fgets(STDIN));
            foreach ($diary['entries'] as &$e) {
                if ($e['id'] == $id) {
                    echo "Новый текст: ";
                    $e['text'] = trim(fgets(STDIN));
                    saveDiary($diary);
                    $lastActivity = time();
                    echo color("✅ Запись обновлена.\n", '32');
                    break 2;
                }
            }
            echo color("❌ Запись не найдена.\n", '31');
            break;
        case '5':
            if (empty($diary['entries'])) {
                echo color("Нет записей.\n", '33');
                break;
            }
            foreach ($diary['entries'] as $e) {
                echo "{$e['id']}: {$e['date']} | {$e['category']} | {$e['text']}\n";
            }
            echo "Введите ID для удаления: ";
            $id = (int) trim(fgets(STDIN));
            foreach ($diary['entries'] as $i => $e) {
                if ($e['id'] == $id) {
                    array_splice($diary['entries'], $i, 1);
                    saveDiary($diary);
                    $lastActivity = time();
                    echo color("✅ Запись удалена.\n", '32');
                    break 2;
                }
            }
            echo color("❌ Запись не найдена.\n", '31');
            break;
        case '6':
            if (empty($diary['entries'])) {
                echo color("Нет записей для экспорта.\n", '33');
                break;
            }
            $fp = fopen('diary_export.csv', 'w');
            fputcsv($fp, ['ID', 'Дата', 'Категория', 'Текст']);
            foreach ($diary['entries'] as $e) {
                fputcsv($fp, [$e['id'], $e['date'], $e['category'], $e['text']]);
            }
            fclose($fp);
            echo color("💾 Экспорт CSV: diary_export.csv\n", '32');
            break;
        case '7':
            if (empty($diary['entries'])) {
                echo color("Нет записей для экспорта.\n", '33');
                break;
            }
            file_put_contents('diary_export.json', json_encode($diary['entries'], JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE));
            echo color("💾 Экспорт JSON: diary_export.json\n", '32');
            break;
        case '8':
            echo "Текущий PIN: ";
            $old = trim(fgets(STDIN));
            echo "Новый PIN (4-6 цифр): ";
            $newPin = trim(fgets(STDIN));
            if (isDigit($newPin) && strlen($newPin) >= 4 && strlen($newPin) <= 6) {
                echo "Повторите новый PIN: ";
                $confirm = trim(fgets(STDIN));
                if ($newPin == $confirm) {
                    if (hashPin($old) == $diary['pin_hash']) {
                        $diary['pin_hash'] = hashPin($newPin);
                        saveDiary($diary);
                        echo color("✅ PIN изменён.\n", '32');
                    } else {
                        echo color("❌ Неверный текущий PIN.\n", '31');
                    }
                } else {
                    echo color("❌ PIN не совпадают.\n", '31');
                }
            } else {
                echo color("❌ PIN должен быть 4-6 цифр.\n", '31');
            }
            break;
        case '9':
            echo "До свидания!\n";
            exit(0);
        default:
            echo color("Неверный выбор.\n", '31');
    }
}
?>
